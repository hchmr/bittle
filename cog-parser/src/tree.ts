export type Point = {
    row: number;
    column: number;
};

export type Range = {
    startIndex: number;
    endIndex: number;
    startPosition: Point;
    endPosition: Point;
};

export interface SyntaxNode {
    tree: Tree;
    id: number;
    type: string;
    isNamed: boolean;
    isMissing: boolean;
    isExtra: boolean;
    hasError: boolean;
    isError: boolean;
    text: string;

    startPosition: Point;
    endPosition: Point;
    startIndex: number;
    endIndex: number;

    parent: SyntaxNode | null;
    children: Array<SyntaxNode>;
    namedChildren: Array<SyntaxNode>;
    childCount: number;
    namedChildCount: number;
    firstChild: SyntaxNode | null;
    firstNamedChild: SyntaxNode | null;
    lastChild: SyntaxNode | null;
    lastNamedChild: SyntaxNode | null;
    nextSibling: SyntaxNode | null;
    nextNamedSibling: SyntaxNode | null;
    previousSibling: SyntaxNode | null;
    previousNamedSibling: SyntaxNode | null;
    descendantCount: number;

    toString(): string;
    child(index: number): SyntaxNode | null;
    namedChild(index: number): SyntaxNode | null;
    childForFieldName(fieldName: string): SyntaxNode | null;
    fieldNameForChild(childIndex: number): string | null;
    childrenForFieldName(fieldName: string): Array<SyntaxNode>;
    firstChildForIndex(index: number): SyntaxNode | null;
    firstNamedChildForIndex(index: number): SyntaxNode | null;

    descendantForIndex(index: number): SyntaxNode;
    descendantForIndex(startIndex: number, endIndex: number): SyntaxNode;
    namedDescendantForIndex(index: number): SyntaxNode;
    namedDescendantForIndex(startIndex: number, endIndex: number): SyntaxNode;
    descendantForPosition(position: Point): SyntaxNode;
    descendantForPosition(startPosition: Point, endPosition: Point): SyntaxNode;
    namedDescendantForPosition(position: Point): SyntaxNode;
    namedDescendantForPosition(startPosition: Point, endPosition: Point): SyntaxNode;
    descendantsOfType(types: string | Array<string>, startPosition?: Point, endPosition?: Point): Array<SyntaxNode>;

    closest(types: string | Array<string>): SyntaxNode | null;
    walk(): TreeCursor;
}

export interface TreeCursor {
    nodeType: string;
    nodeText: string;
    nodeIsNamed: boolean;
    nodeIsMissing: boolean;

    startPosition: Point;
    endPosition: Point;
    startIndex: number;
    endIndex: number;

    readonly currentNode: SyntaxNode;
    readonly currentFieldName: string;
    readonly currentDepth: number;
    readonly currentDescendantIndex: number;

    reset(node: SyntaxNode): void;
    resetTo(cursor: TreeCursor): void;
    gotoParent(): boolean;
    gotoFirstChild(): boolean;
    gotoLastChild(): boolean;
    gotoFirstChildForIndex(goalIndex: number): boolean;
    gotoFirstChildForPosition(goalPosition: Point): boolean;
    gotoNextSibling(): boolean;
    gotoPreviousSibling(): boolean;
    gotoDescendant(goalDescendantIndex: number): void;
}

export interface Tree {
    readonly rootNode: SyntaxNode;
    walk(): TreeCursor;
}
