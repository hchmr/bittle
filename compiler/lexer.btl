include "bittle.btls";

//==============================================================================
//== Helper functions

func try_keyword(lexeme: *Char, keyword: *Char, out: *TokKind, kind: TokKind): Bool {
    if (str_eq(lexeme, keyword)) {
        *out = kind;
        return true;
    } else {
        return false;
    }
}

func lookup_keyword(lexeme: *Char, out: *TokKind): Bool {
    return try_keyword(lexeme, "as", out, Tok_As)
        || try_keyword(lexeme, "break", out, Tok_Break)
        || try_keyword(lexeme, "const", out, Tok_Const)
        || try_keyword(lexeme, "continue", out, Tok_Continue)
        || try_keyword(lexeme, "else", out, Tok_Else)
        || try_keyword(lexeme, "enum", out, Tok_Enum)
        || try_keyword(lexeme, "extern", out, Tok_Extern)
        || try_keyword(lexeme, "false", out, Tok_False)
        || try_keyword(lexeme, "for", out, Tok_For)
        || try_keyword(lexeme, "func", out, Tok_Func)
        || try_keyword(lexeme, "if", out, Tok_If)
        || try_keyword(lexeme, "include", out, Tok_Include)
        || try_keyword(lexeme, "null", out, Tok_Null)
        || try_keyword(lexeme, "return", out, Tok_Return)
        || try_keyword(lexeme, "sizeof", out, Tok_Sizeof)
        || try_keyword(lexeme, "struct", out, Tok_Struct)
        || try_keyword(lexeme, "true", out, Tok_True)
        || try_keyword(lexeme, "var", out, Tok_Var)
        || try_keyword(lexeme, "while", out, Tok_While);
}

//==============================================================================
//== Lexer core

struct Lexer {
    input: *Char,
    index: Int,
    pos: Pos,
    tok_kind: TokKind,
    tok_index: Int,
    tok_pos: Pos,
    tok_lexeme: *Char,
}

func lexer_new(file_name: *Char, input: *Char): *Lexer {
    var self: *Lexer = calloc(1, sizeof(Lexer));

    var pos = Pos { file: file_name, row: 1, col: 1 };

    *self = Lexer {
        input: input,
        index: 0,
        pos: pos,
        tok_kind: Tok_Eof,
        tok_index: 0,
        tok_pos: pos,
        tok_lexeme: "",
    };
    return self;
}

func make_lexeme(self: *Lexer): *Char {
    var len = self.index - self.tok_index;
    var lexeme: *Char = malloc(len + 1);
    memcpy(lexeme, &self.input[self.index - len], len);
    lexeme[len] = '\0';
    return lexeme;
}

func begin_token(self: *Lexer) {
    self.tok_index = self.index;
    self.tok_pos = self.pos;
}

func finish_token(self: *Lexer, tok_kind: TokKind, lexeme: *Char) {
    self.tok_kind = tok_kind;
    self.tok_lexeme = lexeme;
}

func curr_char(self: *Lexer): Int32 {
    return self.input[self.index];
}

func next_char(self: *Lexer) {
    var c = curr_char(self);
    if (c == '\n') {
        self.pos.row += 1;
        self.pos.col = 1;
    } else {
        self.pos.col += 1;
    }
    self.index += 1;
}

//==============================================================================
//== Character matching

func at_end(self: *Lexer): Bool {
    return curr_char(self) == '\0';
}

func at_whitespace(self: *Lexer): Bool {
    return is_whitespace(curr_char(self));
}

func at_word_start(self: *Lexer): Bool {
    return is_word_start(curr_char(self));
}

func at_word_part(self: *Lexer): Bool {
    return is_word_part(curr_char(self));
}

func at_digit(self: *Lexer, base: Int): Bool {
    return is_digit(curr_char(self), base);
}

func at_printable(self: *Lexer): Bool {
    return is_print(curr_char(self));
}

func at_char(self: *Lexer, c: Int32): Bool {
    return curr_char(self) == c;
}

func eat_char(self: *Lexer, c: Int32): Bool {
    if (at_char(self, c)) {
        next_char(self);
        return true;
    } else {
        return false;
    }
}

//==============================================================================
//== Scanning routines

func skip_whitespace(self: *Lexer) {
    while (is_whitespace(curr_char(self))) {
        next_char(self);
    }
}

func skip_line_comment(self: *Lexer) {
    while (!at_end(self) && !eat_char(self, '\n')) {
        next_char(self);
    }
}

func skip_block_comment(self: *Lexer) {
    while (true) {
        if (at_end(self)) {
            die_at(&self.pos, "Unterminated block comment.");
        } else if (eat_char(self, '*') && eat_char(self, '/')) {
            break;
        } else {
            next_char(self);
        }
    }
}

func scan_word(self: *Lexer) {
    next_char(self);
    while (is_word_part(curr_char(self))) {
        next_char(self);
    }
    var lexeme = make_lexeme(self);
    var tok_kind = Tok_Ident;
    lookup_keyword(lexeme, &tok_kind);
    finish_token(self, tok_kind, lexeme);
}

func scan_number(self: *Lexer) {
    var base = 10;
    if (eat_char(self, '0')) {
        if (eat_char(self, 'x')) {
            base = 16;
        } else if (eat_char(self, 'b')) {
            base = 2;
        } else if (eat_char(self, 'o')) {
            base = 8;
        }
    }

    if (base != 10 && !at_digit(self, base)) {
        die_at(&self.pos, "Expected base-%d digit.\n", base);
    }

    while (at_digit(self, base)) {
        next_char(self);
    }

    var lexeme = make_lexeme(self);
    finish_token(self, Tok_Int, lexeme);
}

func scan_char_part(self: *Lexer) {
    if (eat_char(self, '\\')) {
        if (eat_char(self, '\'')) {
        } else if (eat_char(self, '\"')) {
        } else if (eat_char(self, '0')) {
        } else if (eat_char(self, '\\')) {
        } else if (eat_char(self, 'n')) {
        } else if (eat_char(self, 'r')) {
        } else if (eat_char(self, 't')) {
        } else {
            die_at(&self.pos, "Invalid escape sequence.");
        }
    } else if (at_printable(self)) {
        next_char(self);
    } else {
        die_at(&self.pos, "Illegal character.\n");
    }
}

func scan_string(self: *Lexer) {
    next_char(self);
    while (!eat_char(self, '\"')) {
        if (at_end(self)) {
            die_at(&self.pos, "Unterminated string literal.");
        } else {
            scan_char_part(self);
        }
    }
    var lexeme = make_lexeme(self);
    finish_token(self, Tok_String, lexeme);
}

func scan_char(self: *Lexer) {
    next_char(self);
    if (at_char(self, '\'')) {
        die_at(&self.pos, "Empty character literal.");
    }
    scan_char_part(self);
    if (!eat_char(self, '\'')) {
        die_at(&self.pos, "Unterminated character literal.");
    }
    var lexeme = make_lexeme(self);
    finish_token(self, Tok_Char, lexeme);
}

func lexer_next(self: *Lexer): Tok {
    while (true) {
        begin_token(self);
        if (at_end(self)) {
            finish_token(self, Tok_Eof, "");
        } else if (at_whitespace(self)) {
            skip_whitespace(self);
            continue;
        } else if (at_word_start(self)) {
            scan_word(self);
        } else if (at_digit(self, 10)) {
            scan_number(self);
        } else if (at_char(self, '\"')) {
            scan_string(self);
        } else if (at_char(self, '\'')) {
            scan_char(self);
        } else if (eat_char(self, '(')) {
            finish_token(self, Tok_LParen, "(");
        } else if (eat_char(self, ')')) {
            finish_token(self, Tok_RParen, ")");
        } else if (eat_char(self, '[')) {
            finish_token(self, Tok_LBracket, "[");
        } else if (eat_char(self, ']')) {
            finish_token(self, Tok_RBracket, "]");
        } else if (eat_char(self, '{')) {
            finish_token(self, Tok_LBrace, "{");
        } else if (eat_char(self, '}')) {
            finish_token(self, Tok_RBrace, "}");
        } else if (eat_char(self, '+')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_PlusEq, "+=");
            } else {
                finish_token(self, Tok_Plus, "+");
            }
        } else if (eat_char(self, '-')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_MinusEq, "-=");
            } else if (eat_char(self, '>')) {
                finish_token(self, Tok_Arrow, "->");
            } else {
                finish_token(self, Tok_Minus, "-");
            }
        } else if (eat_char(self, '*')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_StarEq, "*=");
            } else {
                finish_token(self, Tok_Star, "*");
            }
        } else if (eat_char(self, '/')) {
            if (eat_char(self, '/')) {
                skip_line_comment(self);
                continue;
            } else if (eat_char(self, '*')) {
                skip_block_comment(self);
                continue;
            } else if (eat_char(self, '=')) {
                finish_token(self, Tok_SlashEq, "/=");
            } else {
                finish_token(self, Tok_Slash, "/");
            }
        } else if (eat_char(self, '%')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_PercentEq, "%=");
            } else {
                finish_token(self, Tok_Percent, "%");
            }
        } else if (eat_char(self, '=')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_EqEq, "==");
            } else {
                finish_token(self, Tok_Eq, "=");
            }
        } else if (eat_char(self, '!')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_BangEq, "!=");
            } else {
                finish_token(self, Tok_Bang, "!");
            }
        } else if (eat_char(self, '<')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_LtEq, "<=");
            } else if (eat_char(self, '<')) {
                if (eat_char(self, '=')) {
                    finish_token(self, Tok_LtLtEq, "<<=");
                } else {
                    finish_token(self, Tok_LtLt, "<<");
                }
            } else {
                finish_token(self, Tok_Lt, "<");
            }
        } else if (eat_char(self, '>')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_GtEq, ">=");
            } else if (eat_char(self, '>')) {
                if (eat_char(self, '=')) {
                    finish_token(self, Tok_GtGtEq, ">>=");
                } else {
                    finish_token(self, Tok_GtGt, ">>");
                }
            } else {
                finish_token(self, Tok_Gt, ">");
            }
        } else if (eat_char(self, '&')) {
            if (eat_char(self, '&')) {
                finish_token(self, Tok_AmpAmp, "&&");
            } else if (eat_char(self, '=')) {
                finish_token(self, Tok_AmpEq, "&=");
            } else {
                finish_token(self, Tok_Amp, "&");
            }
        } else if (eat_char(self, '|')) {
            if (eat_char(self, '|')) {
                finish_token(self, Tok_BarBar, "||");
            } else if (eat_char(self, '=')) {
                finish_token(self, Tok_BarEq, "|=");
            } else {
                finish_token(self, Tok_Bar, "|");
            }
        } else if (eat_char(self, '^')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_CaretEq, "^=");
            } else {
                finish_token(self, Tok_Caret, "^");
            }
        } else if (eat_char(self, '~')) {
            finish_token(self, Tok_Tilde, "~");
        } else if (eat_char(self, '?')) {
            finish_token(self, Tok_Question, "?");
        } else if (eat_char(self, ',')) {
            finish_token(self, Tok_Comma, ",");
        } else if (eat_char(self, ';')) {
            finish_token(self, Tok_Semicolon, ";");
        } else if (eat_char(self, ':')) {
            if (eat_char(self, ':')) {
                finish_token(self, Tok_ColonColon, "::");
            } else {
                finish_token(self, Tok_Colon, ":");
            }
        } else if (eat_char(self, '.')) {
            if (eat_char(self, '.')) {
                if (eat_char(self, '.')) {
                    finish_token(self, Tok_DotDotDot, "...");
                } else {
                    finish_token(self, Tok_DotDot, "..");
                }
            } else {
                finish_token(self, Tok_Dot, ".");
            }
        } else {
            die_at(&self.pos, "Unexpected character: '%c'.\n", curr_char(self));
        }

        return Tok {
            kind: self.tok_kind,
            pos: self.tok_pos,
            lexeme: self.tok_lexeme,
        };
    }
}
