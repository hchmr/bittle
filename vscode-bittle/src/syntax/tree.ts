import { Point } from './position';

export interface SyntaxNode {
    readonly type: string;

    readonly text: string;

    readonly startPosition: Point;
    readonly endPosition: Point;
    readonly startIndex: number;
    readonly endIndex: number;

    readonly tree: Tree;
    readonly parent: SyntaxNode | null;
    readonly children: Array<SyntaxNode>;
    readonly childCount: number;
    readonly firstChild: SyntaxNode | null;
    readonly lastChild: SyntaxNode | null;
    readonly nextSibling: SyntaxNode | null;
    readonly previousSibling: SyntaxNode | null;

    toString(): string;
    pretty(): string;

    child(index: number): SyntaxNode | null;
    childForFieldName(fieldName: string): SyntaxNode | null;
    fieldNameForChild(childIndex: number): string | null;
    childrenForFieldName(fieldName: string): Array<SyntaxNode>;

    descendantForPosition(position: Point): SyntaxNode;
    descendantForPosition(startPosition: Point, endPosition: Point): SyntaxNode;

    descendantsForPosition(position: Point): Array<SyntaxNode>;
    descendantsForPosition(startPosition: Point, endPosition: Point): Array<SyntaxNode>;

    closest(types: string | Array<string>): SyntaxNode | null;
}

export interface Tree {
    readonly text: string;
    readonly rootNode: SyntaxNode;
}
