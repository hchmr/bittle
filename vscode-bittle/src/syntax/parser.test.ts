import { expect, test } from 'vitest';
import { Error, ErrorSink } from './errorSink.js';
import { tokenize } from './lexer.js';
import { Parser } from './parser.js';
import { Tree } from './tree.js';

function parse(text: string): [Tree, Error[]] {
    const errors: Error[] = [];

    const errorSink: ErrorSink = {
        add(error: Error) {
            errors.push(error);
        },
    };

    const tokenizer = tokenize(text, errorSink);

    const parser = new Parser(text, tokenizer, errorSink);

    const tree = parser.top();

    return [tree, errors];
}

const good = `
struct Point {
    x: Int,
    y: Int,
}
func sqr_dst(p1: *Point, p2: *Point): Int {
    var dx = p2.x - p1.x;
    var dy = p2.y - p1.y;
    return dx * dx + dy * dy;
}
`;

const bad = `
func add(a: Int, b: Int): Int {

func maine( {
    print("Helo wolrdn\\n");
`;

test('parser:good', () => {
    const [tree, errors] = parse(good);
    expect(errors).toStrictEqual([]);
});

test('parser:bad', () => {
    const [tree, errors] = parse(bad);
    expect(errors).toMatchSnapshot();
});

test('prettySyntaxTree:good', () => {
    const [tree, _errors] = parse(good);
    expect(tree.rootNode.pretty()).toMatchSnapshot();
});

test('prettySyntaxTree:bad', () => {
    const [tree, _errors] = parse(bad);
    expect(tree.rootNode.pretty()).toMatchSnapshot();
});
