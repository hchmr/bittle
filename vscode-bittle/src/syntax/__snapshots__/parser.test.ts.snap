// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parser:bad 1`] = `
[
  {
    "message": "Unexpected start of statement: 'func'.",
    "position": {
      "column": 0,
      "row": 3,
    },
  },
  {
    "message": "Expected parameter.",
    "position": {
      "column": 12,
      "row": 3,
    },
  },
  {
    "message": "Unexpected start of statement: '<eof>'.",
    "position": {
      "column": 0,
      "row": 5,
    },
  },
]
`;

exports[`prettySyntaxTree:bad 1`] = `
"Root@2:1
  FuncDecl@2:1
    "func"@2:1
    "add"@2:6
    FuncParamList@2:9
      "("@2:9
      FuncParam@2:10
        "a"@2:10
        ":"@2:11
        NameType@2:13
          "Int"@2:13
      ","@2:16
      FuncParam@2:18
        "b"@2:18
        ":"@2:19
        NameType@2:21
          "Int"@2:21
      ")"@2:24
    ":"@2:25
    NameType@2:27
      "Int"@2:27
    BlockStmt@2:31
      "{"@2:31
  FuncDecl@4:1
    "func"@4:1
    "maine"@4:6
    FuncParamList@4:11
      "("@4:11
    BlockStmt@4:13
      "{"@4:13
      ExprStmt@5:5
        CallExpr@5:5
          NameExpr@5:5
            "print"@5:5
          CallArgList@5:10
            "("@5:10
            CallArg@5:11
              LiteralExpr@5:11
                StringLiteral@5:11
                  "\\"Helo wolrdn\\\\n\\""@5:11
            ")"@5:26
        ";"@5:27
      Error@6:1
        ""@6:1
  ""@6:1
"
`;

exports[`prettySyntaxTree:good 1`] = `
"Root@2:1
  StructDecl@2:1
    "struct"@2:1
    "Point"@2:8
    StructBody@2:14
      "{"@2:14
      StructMember@3:5
        "x"@3:5
        ":"@3:6
        NameType@3:8
          "Int"@3:8
      ","@3:11
      StructMember@4:5
        "y"@4:5
        ":"@4:6
        NameType@4:8
          "Int"@4:8
      ","@4:11
      "}"@5:1
  FuncDecl@6:1
    "func"@6:1
    "sqr_dst"@6:6
    FuncParamList@6:13
      "("@6:13
      FuncParam@6:14
        "p1"@6:14
        ":"@6:16
        PointerType@6:18
          "*"@6:18
          NameType@6:19
            "Point"@6:19
      ","@6:24
      FuncParam@6:26
        "p2"@6:26
        ":"@6:28
        PointerType@6:30
          "*"@6:30
          NameType@6:31
            "Point"@6:31
      ")"@6:36
    ":"@6:37
    NameType@6:39
      "Int"@6:39
    BlockStmt@6:43
      "{"@6:43
      LocalDecl@7:5
        "var"@7:5
        "dx"@7:9
        "="@7:12
        BinaryExpr@7:14
          FieldExpr@7:14
            NameExpr@7:14
              "p2"@7:14
            "."@7:16
            "x"@7:17
          "-"@7:19
          FieldExpr@7:21
            NameExpr@7:21
              "p1"@7:21
            "."@7:23
            "x"@7:24
        ";"@7:25
      LocalDecl@8:5
        "var"@8:5
        "dy"@8:9
        "="@8:12
        BinaryExpr@8:14
          FieldExpr@8:14
            NameExpr@8:14
              "p2"@8:14
            "."@8:16
            "y"@8:17
          "-"@8:19
          FieldExpr@8:21
            NameExpr@8:21
              "p1"@8:21
            "."@8:23
            "y"@8:24
        ";"@8:25
      ReturnStmt@9:5
        "return"@9:5
        BinaryExpr@9:12
          BinaryExpr@9:12
            NameExpr@9:12
              "dx"@9:12
            "*"@9:15
            NameExpr@9:17
              "dx"@9:17
          "+"@9:20
          BinaryExpr@9:22
            NameExpr@9:22
              "dy"@9:22
            "*"@9:25
            NameExpr@9:27
              "dy"@9:27
        ";"@9:29
      "}"@10:1
  ""@11:1
"
`;
